### ISSUE TITLE
21 minutes. error waiting for EKS Node Group to create: unexpected state 'CREATE_FAILED'

### ISSUE THREAD

**Issue by rmalenko** [terraform-aws-modules/terraform-aws-eks#1910]
[I use this as example](https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/examples/eks_managed_node_group/main.tf) but I can't reach health node state. Could you prompt me what I did wrong, please?

![image](https://user-images.githubusercontent.com/1563560/156215505-90908a79-1fcb-4294-add9-fb416fbb9272.png)
![image](https://user-images.githubusercontent.com/1563560/156215551-3ac33b71-a149-4865-8888-0423364fd5c2.png)
![image](https://user-images.githubusercontent.com/1563560/156215586-1250e8ad-efe2-4228-b7ac-dc0e41e87bee.png)

**Comment by thedarkwriter** [comment]
I'm hitting the same issues the last few days. My node groups will create but fail. The autoscaling instances will spin up hosts but they won't attach to the cluster. vpc-cni fails to create, etc.
This is my full example:

`
locals {
  node_configs = {
    "dev" = {
      "create" = true
      "min_size" = 2
      "max_size" = 4
      "desired_size" = 2
      "instance_types" = "r5.xlarge"
      "disk_size" = 64
    }
    "prod" = {
      "create" = true
      "min_size" = 2
      "max_size" = 6
      "desired_size" = 4
      "instance_types" = "r5.2xlarge"
      "disk_size" = 128
    }
  }
  node_runner_configs = {
    "dev" = {
      "create" = false
      "min_size" = 0
      "max_size" = 0
      "desired_size" = 0
      "instance_types" = "r5.xlarge"
      "disk_size" = 64
    }
    "prod" = {
      "create" = true
      "min_size" = 2
      "max_size" = 4
      "desired_size" = 4
      "instance_types" = "r5.2xlarge"
      "disk_size" = 128
    }
  }
  environment_node_configs = local.node_configs[var.env]
  environment_node_runner_configs = local.node_runner_configs[var.env]
}

################################################################################
# EKS Module
################################################################################

module "eks" {
  source = "terraform-aws-modules/eks/aws"

  version = "18.7.2"
  cluster_name                    = "NAME"
  cluster_version                 = "1.21"
  cluster_endpoint_private_access = true
  cluster_endpoint_public_access  = false

  cluster_addons = {
    coredns = {
      resolve_conflicts = "OVERWRITE"
    }
    kube-proxy = {}
    vpc-cni = {
      resolve_conflicts = "OVERWRITE"
    }
  }

  cluster_encryption_config = [{
    provider_key_arn = var.kms_key_eks
    resources        = ["secrets"]
  }]

  vpc_id     = var.vpc_id
  subnet_ids = var.private_subnets

  # Extend cluster security group rules
  cluster_security_group_additional_rules = {
    egress_nodes_ephemeral_ports_tcp = {
      description                = "To node 1025-65535"
      protocol                   = "tcp"
      from_port                  = 1025
      to_port                    = 65535
      type                       = "egress"
      source_node_security_group = true
    }
  }

  # Extend node-to-node security group rules
  node_security_group_additional_rules = {
    ingress_self_all = {
      description = "Node to node all ports/protocols"
      protocol    = "-1"
      from_port   = 0
      to_port     = 0
      type        = "ingress"
      self        = true
    }
    egress_all = {
      description      = "Node all egress"
      protocol         = "-1"
      from_port        = 0
      to_port          = 0
      type             = "egress"
      cidr_blocks      = ["0.0.0.0/0"]
      ipv6_cidr_blocks = ["::/0"]
    }
  }

  # EKS Managed Node Group(s)
  eks_managed_node_group_defaults = {
    ami_type               = "AL2_x86_64"
    disk_size              = lookup(local.environment_node_configs, "disk_size") 
    instance_types         = [lookup(local.environment_node_configs, "instance_types")]
    vpc_security_group_ids = [aws_security_group.NAME-eks-nodegroup.id]
  }

  eks_managed_node_groups = {
    wks-eks-121-ng-p1 = {
      create = lookup(local.environment_node_configs, "create")
      min_size = lookup(local.environment_node_configs, "min_size")
      max_size     = lookup(local.environment_node_configs, "max_size")
      desired_size = lookup(local.environment_node_configs, "desired_size")
#      create_iam_role = true
#      create_security_group = true
#      security_group_name = "eks-NAME-ng-121"
#      iam_role_name = "NAME-eks-ng"
      iam_role_attach_cni_policy = false
#      create_launch_template = true
#      launch_template_name = "managed-ng-121"
      #launch_template_name =  aws_launch_template.lt-NAME-eks121-nodegroup.name
      #launch_template_version =  aws_launch_template.lt-NAME-eks121-nodegroup.latest_version
      capacity_type  = "ON_DEMAND"
      key_name = var.default_keypair.key_name
      labels = {
        Environment = var.env
	NodeGroup = "eks-121-ng"
      }


      update_config = {
        max_unavailable_percentage = 50 # or set `max_unavailable`
      }

    }
  }
  depends_on = [
    var.kms_key_eks
  ]
  tags = merge(
    local.common_tags,
    {
      "Cluster" = "NAME"
      "Module" = "eks"	    
    }
  )
}


################################################################################
# Sub-Module Usage on Existing/Separate Cluster
################################################################################

module "eks_managed_node_group" {
  source = "terraform-aws-modules/eks/aws//modules/eks-managed-node-group"

  name            = "wks-eks-121-ng-p2"
  cluster_name    = module.eks.cluster_id
  cluster_version = "1.21"
  vpc_id     = var.vpc_id
  subnet_ids = var.private_subnets
  vpc_security_group_ids = [
    module.eks.cluster_primary_security_group_id,
    module.eks.cluster_security_group_id,
  ]
  tags = merge(
    local.common_tags,
    {
      "Module" = "eks"	    
    }
  )

}



################################################################################
# aws-auth configmap
# Only EKS managed node groups automatically add roles to aws-auth configmap
# so we need to ensure fargate profiles and self-managed node roles are added
################################################################################

data "aws_eks_cluster_auth" "this" {
  name = module.eks.cluster_id
}

locals {
  kubeconfig = yamlencode({
    apiVersion      = "v1"
    kind            = "Config"
    current-context = "terraform"
    clusters = [{
      name = module.eks.cluster_id
      cluster = {
        certificate-authority-data = module.eks.cluster_certificate_authority_data
        server                     = module.eks.cluster_endpoint
      }
    }]
    contexts = [{
      name = "terraform"
      context = {
        cluster = module.eks.cluster_id
        user    = "terraform"
      }
    }]
    users = [{
      name = "terraform"
      user = {
        token = data.aws_eks_cluster_auth.this.token
      }
    }]
  })

  # we have to combine the configmap created by the eks module with the externally created node group/profile sub-modules
  aws_auth_configmap_yaml = <<-EOT
  ${chomp(module.eks.aws_auth_configmap_yaml)}
      - rolearn: ${module.eks_managed_node_group.iam_role_arn}
        username: system:node:{{EC2PrivateDNSName}}
        groups:
          - system:bootstrappers
          - system:nodes
  EOT
}

resource "null_resource" "patch" {
  triggers = {
    kubeconfig = base64encode(local.kubeconfig)
    cmd_patch  = "kubectl patch configmap/aws-auth --patch \"${local.aws_auth_configmap_yaml}\" -n kube-system --kubeconfig <(echo $KUBECONFIG | base64 --decode)"
  }

  provisioner "local-exec" {
    interpreter = ["/bin/bash", "-c"]
    environment = {
      KUBECONFIG = self.triggers.kubeconfig
    }
    command = self.triggers.cmd_patch
  }
}
`

**Comment by aidanmelen** [comment]
Okay; I thought this was just a "me" problem, but now i am fairly certain this is related to the issue that I encounter #1902

**Comment by bryantbiggs** [comment]
@thedarkwriter `iam_role_attach_cni_policy = false` means the VPC CNI plugin cannot attach IPs so nodes won't join your cluster

**Comment by thedarkwriter** [comment]
I'm pretty sure I tried it both ways, but I've switched it and I'll apply it now and report back.

**Comment by bryantbiggs** [comment]
also, there was a slight bug/mis-direction in the EKS managed node group example. I have submitted a fix for it in another PR - you can see the commit [here](https://github.com/terraform-aws-modules/terraform-aws-eks/pull/1905/commits/5d8e6bf85a64b244171e2817c844db2ef17571b4) for more info

tl;dr - you need to have a VPC CNI policy on the nodes when creating a new cluster otherwise the VPC CNI plugin fails to assign IPs and nodes will not join the cluster. It has to be on the nodes since this happens very early in the bootstrap process, we cannot rely on the IRSA role only for VPC CNI unfortunately (only when creating new clusters)

**Comment by thedarkwriter** [comment]
I set this `iam_role_attach_cni_policy = false` in both node group configs.
My security groups are wide open (just in case)
You can see here a sample instance created from the ASG has the EKS_CNI policy
![image](https://user-images.githubusercontent.com/16764577/156242948-f2db4e6d-e694-4517-81a9-cb8b6d1fd075.png)

Then it eventually fails
```shell
│ Error: error waiting for EKS Node Group (wvs:wvs-eks-121-ng-p1-20220301201012804700000011) to create: unexpected state 'CREATE_FAILED', wanted target 'ACTIVE'. last error: 1 error occurred:
│       * i-0963391ebae21684d, i-0c209ec15cdfd0847: NodeCreationFailure: Unhealthy nodes in the kubernetes cluster
│
│
│
│   with module.eks.module.eks.module.eks_managed_node_group["wvs-eks-121-ng-p1"].aws_eks_node_group.this[0],
│   on .terraform/modules/eks.eks/modules/eks-managed-node-group/main.tf line 269, in resource "aws_eks_node_group" "this":
│  269: resource "aws_eks_node_group" "this" {
│
╵
╷
│ Error: error waiting for EKS Node Group (wvs:wvs-eks-121-ng-p2-2022030120100995620000000a) to create: unexpected state 'CREATE_FAILED', wanted target 'ACTIVE'. last error: 1 error occurred:
│       * i-088c08c71b45bc215: NodeCreationFailure: Unhealthy nodes in the kubernetes cluster
│
│
│
│   with module.eks.module.eks_managed_node_group.aws_eks_node_group.this[0],
│   on .terraform/modules/eks.eks_managed_node_group/modules/eks-managed-node-group/main.tf line 269, in resource "aws_eks_node_group" "this":
│  269: resource "aws_eks_node_group" "this" {
```
![image](https://user-images.githubusercontent.com/16764577/156245001-27214dca-df4b-45ba-99cc-5e52d89a464c.png)

I feel like it's a dependency issue - but I'm not sure. I'm spinning up a cluster without he module to see if I run into similar issues. I spent the last day trying different ways to launch a cluster module with the node groups or making the node groups their own modules per the docs but I've varying degrees of bad luck. 

I'm tempted to launch with an older version, such as 17, but it sounds like it is a big lift to refactor to 18 and I want to avoid that.

**Comment by thedarkwriter** [comment]
Well, after some hardcore sleuthing, I found it to be a completely unrelated thing. 
@rmalenko if you are using the default aws/vpc module and created vpc endpoints, and are using a private cluster, you MUST make sure you add a policy for ECR to allow calls. What I found was the VPC endpoints for ecr where blocking all the basic aws-node, kube-proxy and other kube containers which pull from AWS's public ECR repos. If you are using a private cluster with vpc endpoints, make sure you allow 443 in the security group and add this policy:
```json
{
	"Statement": [
		{
			"Action": "ecr:*",
			"Effect": "Allow",
			"Principal": "*",
			"Resource": "*"
		},
		{
			"Action": "ecr:GetAuthorizationToken*",
			"Effect": "Allow",
			"Principal": "*",
			"Resource": "*"
		}
	],
	"Version": "2012-10-17"
}
```

**Comment by bryantbiggs** [comment]
please see the latest example which we have re-verified and added notes clarifying the relationship between CNI policies and IRSA role for VPC CNI
- https://github.com/terraform-aws-modules/terraform-aws-eks/blob/54de3823b69bfe10cfe53fadb97f264bad38b772/examples/eks_managed_node_group/main.tf#L34-L40
- https://github.com/terraform-aws-modules/terraform-aws-eks/blob/54de3823b69bfe10cfe53fadb97f264bad38b772/examples/eks_managed_node_group/main.tf#L99-L104

**Comment by maati55** [comment]
I have similar issue using v18.9.0 

```
module "eks" {
  source                           = "../../../aws-terraform-modules/eks"

  cluster_name                     = "${local.prefix}-eks"
  cluster_version                  = var.cluster_version
  cluster_endpoint_private_access  = true
  cluster_endpoint_public_access   = true
  cluster_enabled_log_types        = ["api", "audit", "authenticator", "controllerManager", "scheduler"]

  create_cni_ipv6_iam_policy = true

  cluster_addons = {
    coredns = {
      resolve_conflicts = "OVERWRITE"
    }
    kube-proxy = {}
    vpc-cni = {
      resolve_conflicts        = "OVERWRITE"
      service_account_role_arn = module.vpc_cni_irsa.iam_role_arn
    }
  }

  cluster_encryption_config = [{
    provider_key_arn = aws_kms_key.eks.arn
    resources        = ["secrets"]
  }]

  vpc_id     = module.vpc.vpc_id
  subnet_ids = module.vpc.private_subnets

  eks_managed_node_group_defaults = {
    ami_type       = "AL2_x86_64"
    disk_size      = 50
    instance_types = ["t3.small"]

    iam_role_attach_cni_policy = true
  }

  eks_managed_node_groups = {
    # Default node group - as provided by AWS EKS
    default_node_group = {
      # By default, the module creates a launch template to ensure tags are propagated to instances, etc.,
      # so we need to disable it to use the default template provided by the AWS EKS managed node group service
      create_launch_template = false
      launch_template_name   = ""

      # Remote access cannot be specified with a launch template
      remote_access = {
        ec2_ssh_key               = aws_key_pair.this.key_name
        source_security_group_ids = [aws_security_group.remote_access.id]
      }
    }
  }

  tags = var.tags
}
```

**Error**:

```
Error: error waiting for EKS Node Group (bds-dev-eks:default_node_group-20220310221721111300000001) to create: unexpected state 'CREATE_FAILED', wanted target 'ACTIVE'. last error: 1 error occurred:
│ 	* i-0f8605772cab93025: NodeCreationFailure: Instances failed to join the kubernetes cluster
│
│
│
│   with module.eks.module.eks_managed_node_group["default_node_group"].aws_eks_node_group.this[0],
│   on ../../../aws-terraform-modules/eks/modules/eks-managed-node-group/main.tf line 269, in resource "aws_eks_node_group" "this":
│  269: resource "aws_eks_node_group" "this" {
│
╵
```

**Comment by anfoxtrot** [comment]
I am having the same issue. @bryantbiggs I'm not sure if I understand the comment correctly.  Are you saying that you must have the configuration
```
create_cni_ipv6_iam_policy = true
...
iam_role_attach_cni_policy = true
```
otherwise the node groups can't join? And then, once the cluster has been created, you can update these values as necessary? 

If so I've tried that and am still receiving the same error.

**Comment by anfoxtrot** [comment]
@thedarkwriter I am also using the `vpc` module and setting up a private cluster.  I tried adding the ECR policies on the node group and its still failing.  Did I misunderstand your solution.  My configuration below

```
module "vpc" {
  source  = "terraform-aws-modules/vpc/aws"
  version = "3.12.0"

  name            = "Infrastructure VPC"
  cidr            = var.vpc_cidr
  azs             = var.vpc_availability_zones
  private_subnets = [var.private_primary_subnet_cidr, var.private_secondary_subnet_cidr]
  public_subnets  = [var.public_primary_subnet_cidr, var.public_secondary_subnet_cidr]

  tags = local.common_tags
}

module "eks" {
  source  = "terraform-aws-modules/eks/aws"
  version = "18.8.1"

  cluster_name                    = var.cluster_name
  cluster_version                 = "1.21"
  cluster_endpoint_private_access = true
  create_cni_ipv6_iam_policy      = true

  vpc_id     = module.vpc.vpc_id
  subnet_ids = concat(module.vpc.private_subnets, [module.vpc.public_subnets[0]])

  cluster_addons = {
    coredns = {
      resolve_conflicts = "OVERWRITE"
    }
    kube-proxy = {}
    vpc-cni = {
      resolve_conflicts        = "OVERWRITE"
      service_account_role_arn = module.vpc_cni_irsa.iam_role_arn
    }
  }

  eks_managed_node_group_defaults = {
    disk_size                    = 20
    instance_types               = ["m5.large"]
    ami_type                     = "AL2_x86_64"
    iam_role_attach_cni_policy   = true
    iam_role_additional_policies = [aws_iam_policy.additional_ecr_access.id]
  }

  eks_managed_node_groups = {
    Infrastructure = {
      subnet_ids   = module.vpc.private_subnets
      min_size     = 2
      max_size     = 7
      desired_size = 2
      labels = {
        Environment = "Infrastructure"
      }
      tags = local.common_tags
    }
  }

  tags = local.common_tags
}

module "vpc_cni_irsa" {
  source  = "terraform-aws-modules/iam/aws//modules/iam-role-for-service-accounts-eks"
  version = "~> 4.12"

  role_name_prefix      = "VPC-CNI-IRSA"
  attach_vpc_cni_policy = true

  oidc_providers = {
    main = {
      provider_arn               = module.eks.oidc_provider_arn
      namespace_service_accounts = ["kube-system:aws-node"]
    }
  }

  tags = local.common_tags
}

resource "aws_iam_policy" "additional_ecr_access" {
  name        = "ecr-read-access"
  description = "Allow ECR access"

  policy = jsonencode({
    Version = "2012-10-17"
    Statement = [
      {
        "Effect" : "Allow",
        "Resource" : "*",
        "Action" : [
          "ecr:GetRegistryPolicy",
          "ecr:DescribeImageScanFindings",
          "ecr:GetLifecyclePolicyPreview",
          "ecr:GetDownloadUrlForLayer",
          "ecr:DescribeRegistry",
          "ecr:DescribePullThroughCacheRules",
          "ecr:DescribeImageReplicationStatus",
          "ecr:GetAuthorizationToken",
          "ecr:ListTagsForResource",
          "ecr:ListImages",
          "ecr:BatchGetRepositoryScanningConfiguration",
          "ecr:GetRegistryScanningConfiguration",
          "ecr:GetAuthorizationToken*",
          "ecr:UntagResource",
          "ecr:BatchGetImage",
          "ecr:DescribeImages",
          "ecr:TagResource",
          "ecr:DescribeRepositories",
          "ecr:BatchCheckLayerAvailability",
          "ecr:GetRepositoryPolicy",
          "ecr:GetLifecyclePolicy"
        ]
      }
    ]
  })

  tags = local.common_tags
}
```

**Comment by sdoyle88** [comment]
@anfoxtrot, I think the issue with your code here is for this module that creates the IRSA, you need to not only specify the **attach_vpc_cni_policy** bool, but you also need to either include the **vpc_cni_enable_ipv4   = true** or **vpc_cni_enable_ipv6   = true** argument.  If you dont do this, then the only permissions that get created in the IRSA role is "CreateTags".  Dug into the IAM module and this is the cause of the issue.  There is a conditional check, not just for **attach_vpc_cni_policy**, but also for the _type of_ policy.

> ```terraform
> module "vpc_cni_irsa" {
>   source  = "terraform-aws-modules/iam/aws//modules/iam-role-for-service-accounts-eks"
>   version = "~> 4.12"
> 
>   role_name_prefix      = "VPC-CNI-IRSA"
>   attach_vpc_cni_policy = true
> ```

I had the same issue when following the example in the EKS module for managed node groups.  I removed the **vpc_cni_enable_ipv6** line because I was using an IPv4 cluster, and noticed the IAM permission policy for the IRSA role was only limited to the CreateTags permission.  This seemed like a conditional check so I dug into the IAM module.  Below is a snippet from the **policies.tf** file in the **iam-role-for-service-accounts-eks** module.   

```hcl
...<omitted for brevity>...
data "aws_iam_policy_document" "vpc_cni" {
  count = var.create_role && var.attach_vpc_cni_policy ? 1 : 0

  # arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy
  dynamic "statement" {
    for_each = var.vpc_cni_enable_ipv4 ? [1] : []
    content {
      sid = "IPV4"
      actions = [
        "ec2:AssignPrivateIpAddresses",
...<omitted for brevity>...
```
Notice that there is conditional check for the attach_vpc_cni_policy at the base level of the data resource, and then another check under the dynamic statement for vpc_cni_enable_ipv4 (there are other checks as well, i.e. vpc_cni_enable_ipv6).  Without this, the permission policy wont have the required permissions to utilize the CNI.   Adding the argument for **vpc_cni_enable_ipv4   = true** successfully allowed my nodes to come online and join the cluster and the IAM permission policy was updated properly with the required Network Interface permissions for the CNI.  

Hope that helps.  Perhaps the example can have a comment in its use of the module to indicate you need to include one versus the other?

**Comment by anfoxtrot** [comment]
@sdoyle88 thanks, that resolved the issue. For anyone that also runs: 

```terraform
module "vpc" {
  source  = "terraform-aws-modules/vpc/aws"
  version = "3.12.0"

  name               = "Infrastructure VPC"
  cidr               = var.vpc_cidr
  azs                = var.vpc_availability_zones
  private_subnets    = [var.private_primary_subnet_cidr, var.private_secondary_subnet_cidr]
  public_subnets     = [var.public_primary_subnet_cidr, var.public_secondary_subnet_cidr]
  enable_nat_gateway = true

}

module "eks" {
  source  = "terraform-aws-modules/eks/aws"
  version = "18.8.1"

  cluster_name                    = var.cluster_name
  cluster_version                 = "1.21"
  cluster_endpoint_private_access = true

  vpc_id     = module.vpc.vpc_id
  subnet_ids = concat(module.vpc.private_subnets, [module.vpc.public_subnets[0]])

  cluster_addons = {
    coredns = {
      resolve_conflicts = "OVERWRITE"
    }
    kube-proxy = {}
    vpc-cni = {
      resolve_conflicts        = "OVERWRITE"
      service_account_role_arn = module.vpc_cni_irsa.iam_role_arn
    }
  }

  eks_managed_node_group_defaults = {
    disk_size                  = 20
    instance_types             = ["m5.large"]
    ami_type                   = "AL2_x86_64"
    iam_role_attach_cni_policy = true
  }

  eks_managed_node_groups = {
    Infrastructure = {
      subnet_ids   = module.vpc.private_subnets
      min_size     = 2
      max_size     = 7
      desired_size = 2
      key_name     = var.key_name
    }
  }
}

module "vpc_cni_irsa" {
  source  = "terraform-aws-modules/iam/aws//modules/iam-role-for-service-accounts-eks"
  version = "~> 4.12"

  role_name_prefix      = "VPC-CNI-IRSA"
  attach_vpc_cni_policy = true
  vpc_cni_enable_ipv4   = true # NOTE: This was what needed to be added

  oidc_providers = {
    main = {
      provider_arn               = module.eks.oidc_provider_arn
      namespace_service_accounts = ["kube-system:aws-node"]
    }
  }
}


```

**Comment by bryantbiggs** [comment]
closing this out - as noted above, at least one of `vpc_cni_enable_ipv4` or `vpc_cni_enable_ipv6` must be enabled to provide the appropriate permissions for the vpc-cni IRSA role

**Comment by gsusI** [comment]
Just in case it helps others: In my case I was using the module `lablabs/eks-cluster-autoscaler/aws`, with `irsa_role_create = false` and `irsa_policy_enabled = false` (assuming that the eks-cluster-autoscaler would use the IAM I set for the `aws_eks_node_group`, however, **I had to provide  the role manually** by using `irsa_assume_role_enabled = true` and `irsa_assume_role_arn=aws_iam_role.kubernetes-nodegroup-iam-role.arn`

**Comment by robsonsantosdevops** [comment]
I was getting the same issue, it was solved by creating another nat gateway and choosing public subnet and attaching that new nat gateway in routable to my private subnet.

**Comment by rupampathak-optimizely** [comment]
> @sdoyle88 thanks, that resolved the issue. For anyone that also runs into this here is what worked for me:
> 
> ```terraform
> module "vpc" {
>   source  = "terraform-aws-modules/vpc/aws"
>   version = "3.12.0"
> 
>   name               = "Infrastructure VPC"
>   cidr               = var.vpc_cidr
>   azs                = var.vpc_availability_zones
>   private_subnets    = [var.private_primary_subnet_cidr, var.private_secondary_subnet_cidr]
>   public_subnets     = [var.public_primary_subnet_cidr, var.public_secondary_subnet_cidr]
>   enable_nat_gateway = true
> 
> }
> 
> module "eks" {
>   source  = "terraform-aws-modules/eks/aws"
>   version = "18.8.1"
> 
>   cluster_name                    = var.cluster_name
>   cluster_version                 = "1.21"
>   cluster_endpoint_private_access = true
> 
>   vpc_id     = module.vpc.vpc_id
>   subnet_ids = concat(module.vpc.private_subnets, [module.vpc.public_subnets[0]])
> 
>   cluster_addons = {
>     coredns = {
>       resolve_conflicts = "OVERWRITE"
>     }
>     kube-proxy = {}
>     vpc-cni = {
>       resolve_conflicts        = "OVERWRITE"
>       service_account_role_arn = module.vpc_cni_irsa.iam_role_arn
>     }
>   }
> 
>   eks_managed_node_group_defaults = {
>     disk_size                  = 20
>     instance_types             = ["m5.large"]
>     ami_type                   = "AL2_x86_64"
>     iam_role_attach_cni_policy = true
>   }
> 
>   eks_managed_node_groups = {
>     Infrastructure = {
>       subnet_ids   = module.vpc.private_subnets
>       min_size     = 2
>       max_size     = 7
>       desired_size = 2
>       key_name     = var.key_name
>     }
>   }
> }
> 
> module "vpc_cni_irsa" {
>   source  = "terraform-aws-modules/iam/aws//modules/iam-role-for-service-accounts-eks"
>   version = "~> 4.12"
> 
>   role_name_prefix      = "VPC-CNI-IRSA"
>   attach_vpc_cni_policy = true
>   vpc_cni_enable_ipv4   = true # NOTE: This was what needed to be added
> 
>   oidc_providers = {
>     main = {
>       provider_arn               = module.eks.oidc_provider_arn
>       namespace_service_accounts = ["kube-system:aws-node"]
>     }
>   }
> }
> ```

I'm still getting the same error, even when copy pasting and trying to apply the above code:

module.eks.module.eks_managed_node_group["Infrastructure"].aws_eks_node_group.this[0]: Still creating... [23m30s elapsed]
╷
│ Error: error waiting for EKS Node Group (airflow-demo:Infrastructure-20221010114123734700000002) to create: unexpected state 'CREATE_FAILED', wanted target 'ACTIVE'. last error: 2 errors occurred:
│       * eks-Infrastructure-20221010114123734700000002-4ec1e0d9-8984-b522-5283-59329b88da2c: AsgInstanceLaunchFailures: Instance became unhealthy while waiting for instance to be in InService state. Termination Reason: Client.InternalError: Client error on launch
│       * i-0145c698cd7fecfa3, i-015fcfc77b9b06212, i-032168ff9c5e2dd68, i-040e153eceeaa8e93, i-044cbbae074d232bc, i-050cefc7d73b2f5c2, i-06ee232f8c88fe5a8, i-0ebd0600958c9c291, i-0f289e3fc86b6e69a, i-0f8349fc1fca311f0: NodeCreationFailure: Instances failed to join the kubernetes cluster
